src/headers.rs

```diff
@@ -0,0 +1,179 @@
+//! This module provides access to the HTTP MIME headers
+//! of a request or response.
+//!
+//! Not to be confused with the request/response metadata
+//! which is also called "headers" in TrafficServer
+
+use crate::bindings::*;
+use crate::helpers::checked_raw_str;
+use std::marker::PhantomData;
+
+/// An iterator over header field values
+///
+/// Skips any invalid strings
+pub struct HeaderValueIter<'a, 'b, T: 'a> {
+    field: &'b HeaderField<'a, 'b, T>,
+    index: usize,
+    length: usize,
+}
+
+impl<'a, 'b, T> Iterator for HeaderValueIter<'a, 'b, T> {
+    type Item = &'a str;
+    fn next(&mut self) -> Option<Self::Item> {
+        while self.index < self.length {
+            let res = checked_raw_str(|len| unsafe {
+                TSMimeHdrFieldValueStringGet(
+                    self.field.headers.bufp,
+                    self.field.headers.offset,
+                    self.field.raw_field,
+                    self.index as i32,
+                    len,
+                )
+            });
+            self.index += 1;
+            if let Ok(res) = res {
+                return Some(res);
+            }
+        }
+        None
+    }
+
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        (0, Some(self.length))
+    }
+}
+
+/// A single MIME header field
+///
+/// Has a name and contains multiple values.
+pub struct HeaderField<'a, 'b, T: 'a> {
+    headers: &'b Headers<'a, T>,
+    raw_field: TSMLoc,
+}
+
+impl<'a, 'b, T> Drop for HeaderField<'a, 'b, T> {
+    fn drop(&mut self) {
+        unsafe { TSHandleMLocRelease(self.headers.bufp, self.headers.offset, self.raw_field) };
```

Comment:
I think if we use rust borrowing we don't need to use the TS refcounting system (which I assume this is here). If we do though, we can omit the borrow and use an owned wrapper like an `Arc`.

EDIT: I think I understand it now. Is this used to release header value fields that were given to Rust? If so, I would be very careful here that any existing `&str` returned by `HeaderValueIter` do not outlive this drop here. I suggest the lifetime of the strings returned from `HeaderValueIter` be bounded on the lifetime of this struct. Even better, the lifetime of the `Headers` struct, and then it would be best, I think, if the `Headers` struct freed _all_ the header values when it is dropped. Then you could iterate over the `Headers` as many times as you want without destroying them.